PubHub/PubHub.AdminPortal/Components/Auth/CustomAuthStateProvider.cs

```diff
@@ -1,20 +1,122 @@
-        private readonly IHttpClientFactory _httpClientFactory;
-        private ClaimsPrincipal _anonymous = new ClaimsPrincipal(new ClaimsIdentity());
+        private readonly IAuthenticationService _authenticationService;
+        private readonly ClaimsPrincipal _anonymous = new(new ClaimsIdentity());
 
-        public CustomAuthStateProvider(ILocalStorageService localStorage, IHttpClientFactory httpClientFactory)
+        public CustomAuthStateProvider(ILocalStorageService localStorage, IAuthenticationService authenticationService)
         {
             _localStorage = localStorage;
-            _httpClientFactory = httpClientFactory;
+            _authenticationService = authenticationService;
+        }
+
+        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
+        {
+            try
+            {
+                var token = await _localStorage.GetItemAsync<string>("token");
+                var refreshToken = await _localStorage.GetItemAsync<string>("refreshToken");
+
+                if (string.IsNullOrWhiteSpace(token) || string.IsNullOrWhiteSpace(refreshToken))
+                    return await Task.FromResult(new AuthenticationState(_anonymous));
+
+                TokenInfo tokenInfo = new() { Token = token, RefreshToken = refreshToken };                
+
+                var claimsPrincipal = GetClaims(token);
+
+                if (claimsPrincipal.FindFirst("exp") != null)
+                {
+                    var expireClaim = long.Parse(claimsPrincipal.FindFirst("exp")?.Value ?? "0");
+
+                    var utcExpire = DateTimeOffset.FromUnixTimeSeconds(expireClaim);
+                    var expiredate = utcExpire.DateTime;
+                    // If the token has expired, then we want to refresh the token and update the existing token and refresh token in our local storage.
+                    if (expiredate < DateTime.UtcNow)
+                    {
+                        var tokenResponse = await _authenticationService.RefreshTokenAsync(tokenInfo);
+                        if (tokenResponse != null && tokenResponse.Instance != null)
+                        {
+                            await _localStorage.SetItemAsync<string>("token", tokenResponse.Instance.Token);
+                            await _localStorage.SetItemAsync<string>("refreshToken", tokenResponse.Instance.RefreshToken);
+                            UpdateAuhenticationState(tokenResponse.Instance.Token);
+                        }
+                    }
+                }
+
+                if (claimsPrincipal.Identity != null && claimsPrincipal.Identity.IsAuthenticated)
+                {
+                    return new AuthenticationState(claimsPrincipal);
+                }
+                else
+                {
+                    // Return an anonymous authentication state if the user is not authenticated
+                    return new AuthenticationState(new ClaimsPrincipal());
```

Comment:
Why does this not follow the `_anonymous` way like the others?

Reply:
Should be changed in the new commits, and we also added something else to the comment below this one; as if it fails to refresh the token, then we will log out the user. I am unsure if that satisfies what you mean in the comment below though. 😊

Reply:
[This](https://github.com/Mike-Mortensen-Portfolio/PubHub_H6_Final/pull/159#discussion_r1555451506) comment?